### PR TITLE
add s8s8 support for quantization on ARM devices

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.
Add support of S8S8 quantization format, with which both activation and weight are quantized in int8 format. This PR adds s8s8 implementation on ARM devices for operators: QLinearConv, QLinearMatMul, QLinearAveragePool, NhwcMaxPool, QLinearGlobalAveragePool. Other quantized operators already have S8S8 support.

QGEMM is the most complex part in the change. Here are the dispatchers for quantized gemm:
|      | ARM64 with dot product           | ARM64                            | ARM32                                    |
|------|----------------------------------|----------------------------------|------------------------------------------|
| U8U8 | MlasGemmU8X8DispatchUdot         | MlasGemmU8X8DispatchNeon         | MlasGemmU8X8DispatchNeon                 |
| U8S8 | MlasGemmU8X8DispatchUdot(B flip) | MlasGemmU8S8DispatchNeon(A flip) | MlasGemmU8X8DispatchNeon(B flip)         |
| S8S8 | MlasGemmS8S8DispatchSdot         | MlasGemmS8S8DispatchNeon         | MlasGemmU8X8DispatchNeon(A flip, B flip) |

### CODE REVIEW

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 1]
```diff
@@ -467,10 +468,23 @@ MlasRequantizeOutput(
             IntegerVector2 = _mm_add_epi32(IntegerVector2, ZeroPointVector);
             IntegerVector3 = _mm_add_epi32(IntegerVector3, ZeroPointVector);
 
-            __m128i WordVector0 = _mm_packus_epi16(IntegerVector0, IntegerVector1);
-            __m128i WordVector1 = _mm_packus_epi16(IntegerVector2, IntegerVector3);
+            __m128i WordVector0;
+            __m128i WordVector1;
+            __m128i ByteVector;
+
+            if (std::is_signed<OutputType>::value) {
```
Comment (zhanghuanrong) [inline onnxruntime/core/mlas/lib/quantize.cpp:475]: if constexpr (...)

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 2]
```diff
@@ -541,8 +555,18 @@ MlasRequantizeOutput(
             IntegerVector = _mm_cvtps_epi32(FloatVector);
             IntegerVector = _mm_add_epi32(IntegerVector, ZeroPointVector);
 
-            IntegerVector = _mm_packus_epi16(IntegerVector, IntegerVector);
-            IntegerVector = _mm_packus_epi16(IntegerVector, IntegerVector);
+            if (std::is_signed<OutputType>::value) {
```
Comment (zhanghuanrong) [inline onnxruntime/core/mlas/lib/quantize.cpp:558]: if constexpr (...) ?

File: onnxruntime/core/mlas/lib/convsym.cpp [hunk 3]
```diff
@@ -191,8 +191,8 @@ MlasConvSymSetOutputZeroPoint(
     bool InputIsSigned
     )
 {
-    constexpr int32_t minimum = InputIsSigned ? std::numeric_limits<int8_t>::min() : std::numeric_limits<uint8_t>::min();
-    constexpr int32_t maximum = InputIsSigned ? std::numeric_limits<int8_t>::max() : std::numeric_limits<uint8_t>::max();
+    int32_t minimum = InputIsSigned ? std::numeric_limits<int8_t>::min() : std::numeric_limits<uint8_t>::min();
```
Comment (zhanghuanrong) [inline onnxruntime/core/mlas/lib/convsym.cpp:194]: different type ? compiler are all ok with such ?

File: onnxruntime/core/mlas/inc/mlas.h [hunk 4]
```diff
@@ -758,7 +758,7 @@ struct MLAS_CONV_SYM_PARAMS {
     const int32_t* Bias;
     const float* Scale;
     bool PerChannelScale;
-    uint8_t OutputZeroPoint;
+    int32_t OutputZeroPoint;
```
Comment (zhanghuanrong) [inline onnxruntime/core/mlas/inc/mlas.h:761]: Should ConvDepthwise() parameters of zeropoint change to int32_t also? (at lines around 706)

File: onnxruntime/core/mlas/lib/convsym.cpp (hunk 3 above)
Comment (yufenglee) [reply onnxruntime/core/mlas/lib/convsym.cpp:194]: yes, it is valid. Here is a list of valid cases: https://en.cppreference.com/w/c/language/operator_other#Conditional_operator

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 5]
```diff
@@ -88,19 +90,26 @@ Status QLinearGlobalAveragePool::Compute(OpKernelContext* context) const {
...
-    case ONNX_NAMESPACE::TensorProto_DataType_UINT8:
-      return ComputeQLinearGlobalAvgPool(X.Data<uint8_t>(), x_scale, *(tensor_x_zero_point->Data<uint8_t>()),
-                                Y.MutableData<uint8_t>(), y_scale, *(tensor_y_zero_point->Data<uint8_t>()),
-                                N, C, image_size, channels_last_, tp);
-    default:
-      ORT_THROW("Unsupported 'dtype' value: ", dtype);
-  }
+  return ComputeQLinearGlobalAvgPool(X.Data<T8Bits>(), x_scale, *(tensor_x_zero_point->Data<T8Bits>()),
+                                     Y.MutableData<T8Bits>(), y_scale, *(tensor_y_zero_point->Data<T8Bits>()),
+                                     N, C, image_size, channels_last_, tp);
```
Comment (chenfucn) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:95]: do you need logic to throw when TensorProto_DataType_FLOAT?

**Comment by chenfucn** [comment]
Thanks for this herculean effort. This is truly a big effort for a critical step toward faster inference in mobile platforms. 

However I think we might have able to break up the changes into smaller and more manageable ones. 
For instance, can we enable s8 average pool op and/or pool op in its own PR?
And since it is guarded with a feature switch that is off by default, it is ok to sacrifice performance for the newly added paths. So can we reuse default C++ kernel for placeholders and add the s8s8 kernel in later pull requests?

 Of course adopting the these suggestions maybe too late for this time. So maybe next time we can be better at small incremental changes that is easier to iterate.

File: onnxruntime/contrib_ops/cpu/quantization/nhwc_max_pool.cc [hunk 6]
```diff
@@ -73,17 +75,17 @@ Status NhwcMaxPool::Compute(OpKernelContext* context) const {
   AllocatorPtr alloc;
   ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
   int64_t col_buffer_batch_count = std::min(output_image_size, output_batch_count);
-  auto* col_data = alloc->Alloc(SafeInt<size_t>(sizeof(const uint8_t*)) * kernel_size * col_buffer_batch_count);
+  auto* col_data = alloc->Alloc(SafeInt<size_t>(sizeof(const T8Bits*)) * kernel_size * col_buffer_batch_count);
   BufferUniquePtr col_buffer(col_data, BufferDeleter(alloc));
-  std::vector<uint8_t> padding_data(static_cast<size_t>(C), 0);
+  std::vector<T8Bits> padding_data(static_cast<size_t>(C), std::numeric_limits<T8Bits>::min());
```
Comment (chenfucn) [inline onnxruntime/contrib_ops/cpu/quantization/nhwc_max_pool.cc:80]: maybe use nmeric_limits<T>::lowest() for consistency between different types?

**Review by zhanghuanrong — DISMISSED** [review]

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 5 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:95]: the schema definition of QLinearGlobalAveragePool can guard the input to be int8 and uint8

**Comment by yufenglee** [comment]
It was split to smaller PRs.

**Pull request closed without merging**